capi-adoption/README.md

```diff
@@ -0,0 +1,114 @@
+
+Along with operators, admission controllers also must support versioning. Due to the nature of webhook architecture and Kubernetes api-server behavior, they introduce an increased risk for cluster lockup because the upstream approach is to only run one webhook at a time on a cluster, but we must ensure that only certain instances of different admission controllers actually do receive the webhook callback. Otherwise there’s risk that a webhook of some older release is crashing on a newer object and blocks the whole operation.
+
+It is also not straightforward to reason about webhooks and how to ensure that some of them only reconcile certain CRs. Currently on Azure we already use upstream CRDs to reconcile workload clusters and therefore azure-admission-controller is defaulting and validating the same CRDs that the upstream webhook would once the clusters are migrated to CAPZ release. It took extensive efforts to find a reliable logic for different label combinations that would work for both cluster flavors because currently the requirements for CR content are very different between Giant Swarm implementations and CAPZ.
+
+When configuring multiple admission controllers to process the same CRs, there is no guaranteed execution order. Currently the primary ordering is implemented with alphabetical sorting, but this is not a documented contract or public API so this could change at any point in the future. Because customers have now full CP API access and it’s even encouraged that they involve themselves with cluster administration, it can happen that a certain combination of CR content is used which then causes Giant Swarm admission controllers to conflict with upstream controllers. Results of these are undefined.
+
+### Versioning of CRDs
+
+Cluster API is still in a very early stage of development and its CRD versions change once or twice a year. Kubernetes api-server can only manage one CRD version at a time as a storage version and backwards and upwards support for alternative versions must be implemented by appropriate webhooks. Depending on the change in a CRD structure these webhooks work or not.
+
+Currently the stable API version for Cluster API is v1alpha3 but v1alpha4 is already well under development. It is highly possible that when we get our first production ready version of GS managed Cluster API released, we soon must migrate to the next one involving CRD migration. This can become very problematic when the management cluster has a large number of active releases that contain components that utilize different versions of CRDs.
```

Comment:
I don't understand this one. This seems to be in relation to the conversion webhook?

Maybe I don't get it but wouldn't this be a problem in any configuration which should be solved by the conversion webhook?

Reply:
The difference between our multi-version situation and upstream situation is following:

When performing in-place upgrade for single set of controllers, involving a CRD upgrade, the process is roughly following:
1. Install conversion webhooks (that provide ability to convert _old -> new_).
2. Update CRD so that old version is still used as a storage version, but new version is served (w/ conversion webhook).
3. "Lock" all CRs from modifications (with simple validating webhook which prevents any changes to CR when given _locked_ annotation is set).
4. Update all controllers. They now read the _old_ CRs as _new_ via conversion webhook.
5. Switch CRD storage version to new one. All CRs are still stored as _old_ because they don't get any changes due to validating webhook.
6. Start removing locks one-by-one from CRs, in order to have controlled upgrade of CRs into new version.

This is one-way train, with a brake lever in case something unexpected would happen. This is also the direction that upstream supports and provides support for.

In our multi-version situation we would need to have conversion webhooks both ways. In case any field gets removed in new version of CRD, we couldn't switch to use it because old operators can't persist the information. Similarly, if we  don't switch to use new version, the added new fields cannot be stored in the CRs and new operators won't work.

Two way conversions when only one can be used as storage are impossible with different sets of changes (and we don't control these changes).

Reply:
I don't quite follow this reasoning. AFAIK cluster-api tests `hub-spoke-hub` so everything I reviewed or was part of implementing (e.g. the ignition changes) needed to have conversion in both directions - from what I remember.

Sure we shouldn't have different versions of operators reconcile the same CR but with `preserveUnknownFields` enabled on the CRD I also don't see how we could lose data. Especially given the fact that `hub-spoke-hub` seems to be one of the things actually tested for.

Ref: See the conversion implemented by Mateusz here which includes both directions https://github.com/kubernetes-sigs/cluster-api/pull/4172/files#diff-65d83b423d8654f175d640d1605e28e0210e47717ea227b4ea1bd83a8800f211R48

Reply:
> with `preserveUnknownFields` enabled on the CRD I also don't see how we could lose data.

I wasn't aware of this behavior (but based my understanding on earlier observations). With that feature the mix&match approach could theoretically work.

Reply:
You can't use `preserveUnknownFields` with v1.CustomResourceDefinition, only v1beta1. For v1, you can add `x-kubernetes-preserve-unknown-fields: true` to the root property of the schema with `type: object` and it will apply to the object and all of its children.

I can't remember the specifics, but Kubernetes does inject an annotation containing the full JSON representation of the resource before it was converted in certain cases. I believe this is intended to avoid data loss in the case of a changed schema.